File: 567-permutation-in-string/permutation-in-string.cpp
class Solution {
public:
    bool checkInclusion(string s1, string s2) {
        int n=s1.length();
        int m=s2.length();

        if(n>m) return false;

        unordered_map<char,int> f_s1;
        unordered_map<char,int> f_s2;
        for(int i=0;i<n;i++)
        {
            f_s1[s1[i]]++;
            f_s2[s2[i]]++;
        }
        if(f_s1==f_s2)
        {
            return true;
        }
        for(int i=n;i<m;i++)
        {
            char chOut=s2[i-n];
            if(f_s2[chOut]==1)
            {
                f_s2.erase(chOut);
            }
            else
            {
                f_s2[chOut]--;
            }
            
            char chToInsert=s2[i];
            f_s2[chToInsert]++;
            if(f_s2==f_s1) return true;
        }
        return false;


    }
};

File: 567-permutation-in-string/permutation-in-string.py
class Solution(object):
    def checkInclusion(self, s1, s2):
        if len(s2) < len(s1):
            return False

        # Initialize the frequency counters
        freq_s1 = Counter(s1)
        freq_s2 = Counter(s2[:len(s1)])

        # Initial comparison
        if freq_s1 == freq_s2:
            return True

        for i in range(len(s1), len(s2)):
            ch_out = s2[i - len(s1)]  # character to remove
            ch_in = s2[i]  # character to add

            # Update the frequency counter for the outgoing character
            if freq_s2[ch_out] == 1:
                del freq_s2[ch_out]
            else:
                freq_s2[ch_out] -= 1

            # Update the frequency counter for the incoming character
            freq_s2[ch_in] += 1

            # Check if the frequency counters are equal
            if freq_s1 == freq_s2:
                return True

        return False
        